### PR TITLE
feat: glightbox types

### PR DESCRIPTION
**Pull request by tomasvn**
- Draft PR for ts types

### CODE REVIEW

**Comment by tomasvn** [comment]
@gingerchew I have update the PR, can you check it out? I am not sure about the typings of methods and setting private fields

**Comment by gingerchew** [comment]
The GLightboxInit needs to be updated since the constructor only takes a config object. As for setting private methods, I think thats fine. Is there something that makes you doubt doing it that way?

**Comment by gingerchew** [comment]
I am no Typescript expert, so I'm interested in your opinion too

**Comment by tomasvn** [comment]
I am not sure about this method `getElementIndex(node: Node) {...}` about the type of the parameter, previously I had nothing, not sure if this is the correct type

**Comment by gingerchew** [comment]
Re: the todo in the last commit, I think this would help https://www.typescriptlang.org/docs/handbook/utility-types.html#returntypetype

```ts
type init = ReturnType<(options: GLightboxConfig) => GLightboxInstance>
```

this could also be a wild goose chase.

File: index.d.ts [hunk 1]
```diff
@@ -519,13 +519,10 @@ declare class GlightboxInit {
...
 
-/**
- * TODO
- * How to type this default function?
- * When trying to click imported file it goes to this exported function not the Options
- */
-export default function (options = {}) {
-    const instance = new GlightboxInit(options);
+export default function (options: Glightbox.Options) {
+    let instance: InstanceType<typeof GlightboxInit>;
```
Comment (gingerchew) [inline index.d.ts:523]: Can this not be written as:
```ts
export default function(options: GLightbox.Options) {
    const instance:InstanceType<typeof GlightboxInit> = new GlightboxInit(otions);
    instance.init();

    return instance;
}
```
Or was does that not work in this case?

Comment (tomasvn) [reply index.d.ts:523]: @gingerchew the ts compiler returns an error if I write it as `const instance = new Glightbox({...})`

```
const' initializer in an ambient context must be a string or numeric literal" and "(TS) Initializers are not allowed in ambient contexts.
```

Comment (tomasvn) [reply index.d.ts]: I have moved away from initializing the class, and typing the instance instead

Comment (tomasvn) [reply index.d.ts]: @gingerchew I have improved the types, now the intellsense is much better, can you review last commit 👍 

image

**Review by gingerchew — CHANGES_REQUESTED** [review]
Edit: This is referring to [the improve trigger method commit](https://github.com/biati-digital/glightbox/pull/411/commits/674a0a959bef7fa90b268acad2540396b4f9c022)

This should have a type for data shouldn't it?

Looking at the `showSlide` method, [here on line 276](https://github.com/biati-digital/glightbox/blob/32087845a3872bc875139af9a0a66ae6f92b2240/src/js/glightbox.js#L276) and on [line 282](https://github.com/biati-digital/glightbox/blob/32087845a3872bc875139af9a0a66ae6f92b2240/src/js/glightbox.js#L282) It looks like at least the slideData is passed on for some events. There's also this one inside of the `preloadSlide` method [on line 438](https://github.com/biati-digital/glightbox/blob/32087845a3872bc875139af9a0a66ae6f92b2240/src/js/glightbox.js#L438-L446)

```js
        this.trigger('slide_inserted', {
            index: index,
            slide: addedSlideNode,
            slideNode: addedSlideNode,
            slideConfig: data,
            slideIndex: index,
            trigger: null,
            player: addedSlidePlayer
        });
```

Then in `removeSlide` [on line 482](https://github.com/biati-digital/glightbox/blob/32087845a3872bc875139af9a0a66ae6f92b2240/src/js/glightbox.js#L482) the data is just a slide index (number).

Inside `slideAnimateIn` the `slide_changed` event is triggered with *yet another* data object being passed. This is [on line 529](https://github.com/biati-digital/glightbox/blob/32087845a3872bc875139af9a0a66ae6f92b2240/src/js/glightbox.js#L529-L532) and [on line 552](https://github.com/biati-digital/glightbox/blob/32087845a3872bc875139af9a0a66ae6f92b2240/src/js/glightbox.js#L552-L555).

In `slideAnimateOut` there is this monstrosity [on line 587](https://github.com/biati-digital/glightbox/blob/32087845a3872bc875139af9a0a66ae6f92b2240/src/js/glightbox.js#L587-L606).

```js
        this.trigger('slide_before_change', {
            prev: {
                index: this.prevActiveSlideIndex, //this will be removed in the future
                slide: this.prevActiveSlide, //this will be removed in the future
                slideNode: this.prevActiveSlide,
                slideIndex: this.prevActiveSlideIndex,
                slideConfig: _.isNil(this.prevActiveSlideIndex) ? null : this.elements[this.prevActiveSlideIndex].slideConfig,
                trigger: _.isNil(this.prevActiveSlideIndex) ? null : this.elements[this.prevActiveSlideIndex].node,
                player: this.getSlidePlayerInstance(this.prevActiveSlideIndex)
            },
            current: {
                index: this.index, //this will be removed in the future
                slide: this.activeSlide, //this will be removed in the future
                slideNode: this.activeSlide,
                slideIndex: this.index,
                slideConfig: this.elements[this.index].slideConfig,
                trigger: this.elements[this.index].node,
                player: this.getSlidePlayerInstance(this.index)
            }
        });
```

I wish I had a suggestion other than "*Unions?*" but I don't think I've run into something like this before when using typescript.

**Comment by tomasvn** [comment]
I have improved upon your suggestions, yeah haven't found other solution rarther than to use unions I have to think about, how it can be improved

**Comment by tomasvn** [comment]
Had to go to reddit, not going to lie :D Well I had to reach out to better TS devs

**Comment by gingerchew** [comment]
Even still, great find!

I was looking through and found a place where the function arguments don't match up. Here [on line 340](https://github.com/biati-digital/glightbox/blob/0e0dd79cf8452e9f1f84870584b2417a8d45f7f0/index.d.ts#L340-L348) of the `d.ts` file, the `open` and `openAt` methods use the same arguments.

`open` should have an element and index arguments. While I think this would work:
```ts
open(element:Element, index:number)
```

both have default parameters of `null`, will typing it like above mess with anything? 
```js
open(element = null, startAt = null)
```

**Comment by tomasvn** [comment]
> Even still, great find!
> 
> I was looking through and found a place where the function arguments don't match up. Here [on line 340](https://github.com/biati-digital/glightbox/blob/0e0dd79cf8452e9f1f84870584b2417a8d45f7f0/index.d.ts#L340-L348) of the `d.ts` file, the `open` and `openAt` methods use the same arguments.
> 
> `open` should have an element and index arguments. While I think this would work:
> 
> ```ts
> open(element:Element, index:number)
> ```
> 
> both have default parameters of `null`, will typing it like above mess with anything?
> 
> ```js
> open(element = null, startAt = null)
> ```

Nice find, must have missed this one, no it should be fine

**Comment by JorgeMorata** [comment]
Hi, when will be available this PR?? Thank u!

**Comment by gingerchew** [comment]
The existing changes are probably safe to copy and paste into your own `.d.ts` file. There is more stuff internally that needs some additional typing before it makes sense to merge it in.

**Comment by biati-digital** [comment]
Hi everyone. Apologies for the delay. I've been working on v4, which got a total overhaul. Created a plugins API, revamped the website, docs, and more.

I'll be merging a few PRs and dropping the final update in the next days.

**Pull request merged**